### PR TITLE
New lint for needless use of nightly features

### PR DESCRIPTION
**Pull request by skade**
This is still work in progress, but my first lint :). It detects the needless use of nightly features that have questionable chances of making it to stable or have stable equivalents nowadays.

Before I continue working on this, I have 2 questions:
* Is such a lint within the scope of clippy?
* Should the lint also check whether the feature is really opted in or will clippy just release a new version should a feature become stable?

### CODE REVIEW

**Comment by llogiq** [comment]
Hi @skade, cool to see you joining in! :smile:

As for your questions:

* Yes, it is *at least* within the scope of clippy. I think we may even want to move it into rustc if it proves to work well.
* I don't think that unstable features *can* be used unless they have been opted in, unless there is a bug in rustc I don't know of. Regarding the stability of features, we may want to track rust development anyway (i.e. adding more lints for unstable features as we extend this, removing lints that target newly stabilized features).

**Comment by Manishearth** [comment]
I agree that this is within the scope of clippy.

> I think we may even want to move it into rustc if it proves to work well.

Note that there's very little chance of of lints being upstreamed to rustc right now, Alex has mentioned that he wants less in-tree lints, not more.

However, a tag on `#[unstable()]` that suggests alternatives would be nice. It wouldn't help people who have already opted in though.

> I don't think that unstable features can be used unless they have been opted in

The idea behind this lint is to move people off of nightly features wherever possible.

**Comment by llogiq** [comment]
> Note that there's very little chance of of lints being upstreamed to rustc right now, Alex has mentioned that he wants less in-tree lints, not more.

I know. I still think it would be useful to discourage more people from using nightly who don't strictly need it, and the rustc userbase is much larger than clippy's.

> The idea behind this lint is to move people off of nightly features wherever possible.

Full ack, I just meant to say we don't need to check for the opt-in. If the user hasn't opted in, rustc won't allow them to use the feature anyway.

**Comment by Manishearth** [comment]
Oh, I see.

The check for the opt in is so that clippy can detect if the feature became stable (if there is no flag, then it must be stable) and not warn in that case.

I think a list of available features is available somewhere in the LintContext, I'll find it when I'm back on a laptop.

**Comment by llogiq** [comment]
@Manishearth you probably think of `Session::features` which is a `syntax::feature_gate::Features` and has a good number of booleans for unstable language features plus
`declared_stable_lang_features: Vec<Span>` and `declared_lib_features: Vec<(InternedString, Span)>`.

**Comment by skade** [comment]
I didn't explain the case for a check for an opt-in well enough.

* `as_slice()` is unstable. Lint complains, but doesn't check if the user used `#![feature]`
* Through some weird event, `as_slice()` suddenly finds a cult following and is adopted.
* Clippy still warns about `as_slice()` until we figure out and patch clippy to drop the lint.

As users have to opt-in if it is a feature, checking for that would at least ensure that people don't run into stray messages. Not sure if thats a huge effort for not much gain, though.

**Comment by skade** [comment]
*shakes fist at rust-nightly*

I think given that clippy breaks regularly because of rust nightly means that we can also follow their feature changes a little :).

**Comment by llogiq** [comment]
Quod erat demonstrandum.

File: src/needless_features.rs [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+                           instead");
+            }
+            if name.node.as_str() == "as_mut_slice" {
+                span_lint(cx, AS_MUT_SLICE, expr.span,
+                          "used as_mut_slice() from the 'convert' nightly feature. Use &mut [..] \
+                           instead");
+            }
+        }
+    }
+}
```
Comment (Manishearth) [inline src/needless_features.rs]: Nit: newline at end of file

File: src/needless_features.rs [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+impl LintPass for NeedlessFeaturesPass {
+    fn get_lints(&self) -> LintArray {
+        lint_array!(AS_SLICE,AS_MUT_SLICE)
+    }
+}
+
+impl LateLintPass for NeedlessFeaturesPass {
+    fn check_expr(&mut self, cx: &LateContext, expr: &Expr) {
+        if let ExprMethodCall(ref name, _, _) = expr.node {
+            if name.node.as_str() == "as_slice" {
```
Comment (Manishearth) [inline src/needless_features.rs]: This has a false positive with other methods of the same name. We should use `match_def_path` here on the method call node ID and verify that the path is `collections::vec::Vec::as_slice`, or whatever.

File: src/needless_features.rs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+//!
+//! This lint is **warn** by default
+
+use rustc::lint::*;
+use rustc_front::hir::*;
+
+use utils::{span_lint};
+
+declare_lint! {
+    pub AS_SLICE,
```
Comment (Manishearth) [inline src/needless_features.rs]: :bike: :house: Perhaps we should call these `unstable_as_slice`, etc?

File: src/needless_features.rs (hunk 2 above)
Comment (skade) [inline src/needless_features.rs]: I was checking that, all other `as_slice` in std are covered by `convert`. Still, people might write one of their own. The final decision whether something is covered by a feature is still done by the compiler. (can I access that somewhere?)

Comment (Manishearth) [inline src/needless_features.rs]: We should then match against all of those :)

You can also use [`get_attrs`](http://manishearth.github.io/rust-internals-docs/rustc/middle/ty/struct.ctxt.html#method.get_attrs) (after getting the defid via  `cx.tcx.def_map`) and check if the unstable attribute is there.

File: tests/compile-fail/needless_features.rs [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+
+#![deny(clippy)]
+
+fn test_as_slice() {
+    let v = vec![1];
+    v.as_slice(); //~ERROR used as_slice() from the 'convert' nightly feature. Use &[..]
+
+    let mut v2 = vec![1];
+    v2.as_mut_slice(); //~ERROR used as_mut_slice() from the 'convert' nightly feature. Use &mut [..]
+}
```
Comment (llogiq) [inline tests/compile-fail/needless_features.rs]: A test against false positives would be good – something like:

```
struct ShouldWork;

impl ShouldWork {
    fn as_slice(&self, index: usize) -> &ShouldWork { self }
}

fn test_should_work() {
    let sw = ShouldWork;
    sw.as_slice(0);
}
```

**Comment by llogiq** [comment]
If this is just about library features (e.g. methods/functions), I think the `#[unstable]` attribute should be available from the std rlib, so we could get the attributes from the method and thus find all calls to unstable methods. However, we'd still need to match the feature flag to get good lint messages.

**Comment by Manishearth** [comment]
We only want all calls to unstable methods which have a reasonable stable equivalent.

**Comment by llogiq** [comment]
That's why I suggested matching the feature flag name.

**Comment by skade** [comment]
I'll add a little running list of additional lints I want here. Feel free to suggest:

*[ ] Benchmarks in main crate code. Use the `benches` folder and `cargo bench` instead.

**Comment by llogiq** [comment]
Please open another issue for each lint suggestion.

**Comment by skade** [comment]
Ah, okay, will do that. I had considered the lint module as the unit of granularity :).

**Comment by skade** [comment]
I hope I caught all of the comments. I validated that Vec::as_slice() is indeed the only location where the methods are not already _deprecated_.

**Comment by llogiq** [comment]
Needs another `util/update_lints.py` call.

**Pull request merged**